### PR TITLE
fix long wait for process exit in ShellCommand

### LINKED ISSUE #38889: ArrowStream feeding into executable has an unexpected slowdown of ~1 sec
Playing with various formats for the `executable()` table function got to a peculiar perf problem.
Query:
```sql
caspiandb6 :) SELECT * FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (SELECT * FROM numbers(10*1000000)));

SELECT *
FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (
    SELECT *
    FROM numbers(10 * 1000000)
))

Query id: b73e5936-22d1-4d4b-826e-18d3aab80e47

┌────────────sum─┐
│ 49999995000000 │
└────────────────┘

1 row in set. Elapsed: 1.131 sec. 

```
1.131 sec ? seems slow as hell
Let's decompose the query:
```bash
$ time ./clickhouse-client --query="SELECT * FROM numbers(10*1000000) FORMAT ArrowStream" > /tmp/file

real	0m0.097s
user	0m0.019s
sys	0m0.065s

$ time cat /tmp/file | ./user_scripts/arrow_sum.py >/dev/null

real	0m0.128s
user	0m0.182s
sys	0m0.301s
```
0.097 + 0.128 = 0.225 sec
This looks about right.

Let's try different array sizes:
```sql
caspiandb6 :) SELECT * FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (SELECT * FROM numbers(1*1000000)));

SELECT *
FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (
    SELECT *
    FROM numbers(1 * 1000000)
))

Query id: 26965ad9-8115-4b59-a0cd-c062a7e0658a

┌──────────sum─┐
│ 499999500000 │
└──────────────┘

1 row in set. Elapsed: 1.094 sec. 

caspiandb6 :) SELECT * FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (SELECT * FROM numbers(100*1000000)));

SELECT *
FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (
    SELECT *
    FROM numbers(100 * 1000000)
))

Query id: e3dea8a3-1fb6-4d25-9182-1d9aadd12e7b

┌──────────────sum─┐
│ 4999999950000000 │
└──────────────────┘

1 row in set. Elapsed: 1.302 sec. 
```
It looks like 1 sec is just added to all queries.
Seems like a famous `sleep 1` problem somewhere in the code, frankly. :)
@kitaisreal any idea what could be the problem?
I will continue to poke it meanwhile.

For the sake of completeness:
`arrow_sum.py`
```python
#!/usr/bin/python3

import sys
import traceback
import pyarrow as pa
import pyarrow.compute as pc

if __name__ == "__main__":
    fd = open("/tmp/debug.log", "w")
    sum = pa.scalar(0, type="UInt64")
    try:
        with pa.ipc.open_stream(sys.stdin.buffer) as reader:
            for b in reader:
                sum = pc.add(sum, pc.sum(b[0]))
        array = pa.array([sum.as_py()], type="UInt64")
        out_batch = pa.record_batch([array], names=['sum'])
        with pa.ipc.new_stream(sys.stdout.buffer, out_batch.schema) as writer:
            writer.write(out_batch)
    except Exception as e:
        traceback.print_exc(file=fd)
        fd.close()
    sys.stdout.flush()

```

### PR DESCRIPTION
**Pull request by pkit**
fixes #38889

### Changelog category:
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Executable UDF, Executable Dictionary, Executable Storage poll subprocess fix 1 second subprocess wait during subprocess termination

### CODE REVIEW

**Comment by pkit** [comment]
@kitaisreal fixed OSX code, should actually work

File: src/Common/waitForPid.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        int waitpid_res = waitpid(pid, &status, WNOHANG);
+        bool process_terminated_normally = (waitpid_res == pid);
+        return process_terminated_normally;
+    }
+
+    /// If timeout is positive try waitpid without block in loop until
+    /// process is normally terminated or waitpid return error
+
+    int timeout_in_ms = timeout_in_seconds * 1000;
```
Comment (kitaisreal) [inline src/Common/waitForPid.cpp:154]: I would prefer that this here we just call return `pollPid(pid, timeout_in_ms)`. And `pollPid` will handle EINTR, and other stuff.

Comment (kitaisreal) [reply src/Common/waitForPid.cpp:154]: Main motivation is that checking `errno` in caller could lead to different issues. For example in `pollPid` we close descriptor in last line, but later in caller we check `errno` probably assuming that this it come from `poll`. 
```
int ready = poll(&pollfd, 1, timeout_in_ms);
close(pid_fd);
return ready;
```

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Common/waitForPid.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        }
+        else if (waitpid_res == 0)
+        {
+            watch.start();
+            int ready = pollPid(pid, timeout_in_ms);
+            if (ready <= 0)
+            {
+                if (errno == EINTR || errno == EAGAIN)
+                {
+                    timeout_in_ms -= watch.elapsedMilliseconds();
```
[the remark this answers is not in the document]
Comment (pkit) [reply src/Common/waitForPid.cpp:171]: I restart it with `watch.start()` above.

[the remark this answers is not in the document]
Comment (pkit) [reply src/Common/waitForPid.cpp:171]: I can use `watch.restart()` which is just an alias to `watch.start()`

File: src/Common/waitForPid.cpp (hunk 1 above)
Comment (pkit) [reply src/Common/waitForPid.cpp:158]: Okay, saving-restoring errno seems to work fine, and it should be thread-safe

**Comment by pkit** [comment]
@kitaisreal updated :)

**Review by kitaisreal — APPROVED** [review]

**Comment by kitaisreal** [comment]
Need to fix build https://s3.amazonaws.com/clickhouse-builds/38929/2aab66d21ab9f8429c3d7c929aa6e34604e73868/binary_freebsd/build_log.log.
And style check https://s3.amazonaws.com/clickhouse-test-reports/38929/2aab66d21ab9f8429c3d7c929aa6e34604e73868/style_check__actions_.html.

**Comment by pkit** [comment]
@kitaisreal fixed. Seems like FastTest is stuck for 1 hour waiting for a runner though. :)

**Pull request merged**